### PR TITLE
Cache instantiation expression types early enough to prevent reentrancy during printback

### LINKED ISSUE #59821: RangeError: Maximum call stack size exceeded
### 🔎 Search Terms

RangeError: Maximum call stack size exceeded

### 🕗 Version & Regression Information

- This happens in all versions I tried - from 4.9.5 to 5.5.4
- This statred in commit [ae415](https://github.com/gentlee/react-redux-cache/commit/ae4159b12ab02110874ffda5383ea8605b5601e7) in [my repo](https://github.com/gentlee/react-redux-cache)
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about it
- I was unable to test this on prior versions because they are too old

### ⏯ Playground Link

https://github.com/gentlee/react-redux-cache/commit/ae4159b12ab02110874ffda5383ea8605b5601e7

### 💻 Code

Probably caused by cycled typing of `createCacheReducer -> Cache` and `Cache -> createCacheReducer` in `src/reducer`:
```ts
export const createCacheReducer = <N extends string, T extends Typenames, QP, QR, MP, MR>(
   actions: ReturnType<typeof createActions<N, T, QR, MR>>,
   typenames: Cache<N, T, QP, QR, MP, MR>['typenames'],
   queries: Cache<N, T, QP, QR, MP, MR>['queries'],
   mutations: Cache<N, T, QP, QR, MP, MR>['mutations'],
   cacheOptions: Cache<N, T, QP, QR, MP, MR>['options']
   ...
```

### 🙁 Actual behavior

`yarn tsc` throws:

```
RangeError: Maximum call stack size exceeded
    at typeToTypeNodeWorker (typescript/lib/tsc.js:44218:42)
    at typeToTypeNodeHelper (/typescript/lib/tsc.js:44214:32)
    at typeToTypeNodeWorker (ypescript/lib/tsc.js:44438:42)
    at typeToTypeNodeHelper (typescript/lib/tsc.js:44214:32)
    at mapToTypeNodes (typescript/lib/tsc.js:44974:40)
    at typeToTypeNodeWorker (typescript/lib/tsc.js:44355:45)
    at typeToTypeNodeHelper (typescript/lib/tsc.js:44214:32)
    at mapToTypeNodes (typescript/lib/tsc.js:44974:40)
    at typeToTypeNodeWorker (typescript/lib/tsc.js:44355:45)
    at typeToTypeNodeHelper (typescript/lib/tsc.js:44214:32)
```

### 🙂 Expected behavior

Should be without errors.

### Additional information about the issue

In `main` branch of the repo problem is not happening.

it started happening in `support-multiple-caches` branch (1 commit ahead) mostly after code from `createActions` file was added and used.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #59821

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -37400,9 +37400,17 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
+        if (links.instantiationExpressionTypes.has(exprType.id)) {
+            return links.instantiationExpressionTypes.get(exprType.id)!;
+        }
         let hasSomeApplicableSignature = false;
         let nonApplicableType: Type | undefined;
         const result = getInstantiatedType(exprType);
+        links.instantiationExpressionTypes.set(exprType.id, result);
         const errorType = hasSomeApplicableSignature ? nonApplicableType : exprType;
         if (errorType) {
             diagnostics.add(createDiagnosticForNodeArray(getSourceFileOfNode(node), typeArguments, Diagnostics.Type_0_has_no_signatures_for_which_the_type_argument_list_is_applicable, typeToString(errorType)));
```
Comment (weswigham) [inline src/compiler/checker.ts:37416]: The root cause of the stack overflow was basically that the `typeToString` on this line eventually recursively calls into `getInstantiationExpressionType` to make the version of the type that's a member of the mapped type being printed, which is identical, but re-triggers the error emit and then spins.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -37400,9 +37400,17 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (exprType === silentNeverType || isErrorType(exprType) || !some(typeArguments)) {
             return exprType;
         }
+        const links = getNodeLinks(node);
+        if (!links.instantiationExpressionTypes) {
+            links.instantiationExpressionTypes = new Map();
+        }
+        if (links.instantiationExpressionTypes.has(exprType.id)) {
+            return links.instantiationExpressionTypes.get(exprType.id)!;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:37408]:
```suggestion
        links.instantiationExpressionTypes ??= new Map();

        const cachedType = links.instantiationExpressionTypes.get(exprType.id);
        if (cachedType) {
            return cachedType;
```

File: src/compiler/types.ts [hunk 3]
```diff
@@ -6238,6 +6238,7 @@ export interface NodeLinks {
     potentialReflectCollisions?: Node[];
     potentialUnusedRenamedBindingElementsInTypes?: BindingElement[];
     externalHelpersModule?: Symbol;     // Resolved symbol for the external helpers module
+    instantiationExpressionTypes?: Map<number, Type>; // Cache of instantiation expression types for the node
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts:6241]: Is there any reason you cache by IDs instead of by Type?

Comment (weswigham) [reply src/compiler/types.ts:6241]: Historic convention and habit, mostly. I can swap it if you'd like it.

**Review by sandersn — APPROVED** [review]

**Pull request merged**